core/serialize/serialize.cc

```diff
@@ -1035,19 +981,90 @@ void SerializerImpl::unpickleGS(UnPickler &p, GlobalState &result) {
         Timer timeit(result.tracer(), "moving");
         result.fileRefByPath = std::move(fileRefByPath);
         result.files = std::move(files);
-        result.utf8Names = std::move(utf8Names);
-        result.constantNames = std::move(constantNames);
-        result.uniqueNames = std::move(uniqueNames);
         result.classAndModules = std::move(classAndModules);
         result.methods = std::move(methods);
         result.fields = std::move(fields);
         result.typeArguments = std::move(typeArguments);
         result.typeMembers = std::move(typeMembers);
-        result.namesByHash = std::move(namesByHash);
     }
     result.sanityCheck();
 }
 
+void SerializerImpl::pickleNameTable(Pickler &p, const GlobalState &gs) {
+    p.putU4(gs.utf8Names.size());
+    for (const auto &n : gs.utf8Names) {
+        pickle(p, n);
+    }
+    p.putU4(gs.constantNames.size());
+    for (const auto &n : gs.constantNames) {
+        pickle(p, n);
+    }
+    p.putU4(gs.uniqueNames.size());
+    for (const auto &n : gs.uniqueNames) {
+        pickle(p, n);
+    }
+
+    p.putU4(gs.namesByHash.size());
+    for (const auto &s : gs.namesByHash) {
+        p.putU4(s.hash);
+        p.putU4(s.rawId);
+    }
+}
+
+void SerializerImpl::unpickleNameTable(UnPickler &p, GlobalState &result) {
+    vector<UTF8Name> utf8Names(std::move(result.utf8Names));
+    utf8Names.clear();
+    vector<ConstantName> constantNames(std::move(result.constantNames));
+    constantNames.clear();
+    vector<UniqueName> uniqueNames(std::move(result.uniqueNames));
+    uniqueNames.clear();
+    vector<GlobalState::Bucket> namesByHash(std::move(result.namesByHash));
+    namesByHash.clear();
+
+    {
+        Timer timeit(result.tracer(), "readNames");
+
+        int namesSize = p.getU4();
+        ENFORCE_NO_TIMER(namesSize > 0);
+        utf8Names.reserve(nextPowerOfTwo(namesSize));
+        for (int i = 0; i < namesSize; i++) {
+            utf8Names.emplace_back(unpickleUTF8Name(p, result));
+        }
+        namesSize = p.getU4();
+        ENFORCE_NO_TIMER(namesSize > 0);
+        constantNames.reserve(nextPowerOfTwo(namesSize));
+        for (int i = 0; i < namesSize; i++) {
+            constantNames.emplace_back(unpickleConstantName(p, result));
+        }
+        namesSize = p.getU4();
+        ENFORCE_NO_TIMER(namesSize > 0);
+        uniqueNames.reserve(nextPowerOfTwo(namesSize));
+        for (int i = 0; i < namesSize; i++) {
+            uniqueNames.emplace_back(unpickleUniqueName(p, result));
+        }
+    }
+
+    {
+        Timer timeit(result.tracer(), "readNameTable");
+        int namesByHashSize = p.getU4();
+        namesByHash.reserve(namesByHashSize);
+        for (int i = 0; i < namesByHashSize; i++) {
+            auto hash = p.getU4();
+            auto value = p.getU4();
+            namesByHash.emplace_back(GlobalState::Bucket{hash, value});
+        }
+    }
+
+    {
+        Timer timeit(result.tracer(), "moveNames");
```

Comment:
Is there a reason to suspect that this block is ever slow?

Reply:
Nope, I think the moves should take no time. Again, I was preserving the existing structure, but am happy to clean this up while I'm here.